0009-comit-btsieve-db.adoc

```diff
@@ -0,0 +1,245 @@
+
+=== paenkoDB
+
+Type: `Document` +
+https://github.com/paenko/paenkodb
+
+- - unmainted
+
+== WASM consideration
+
+The decision of this spike does not affect any future WASM plans since we will have to provide a different storage layer implementation anyway.
+WASM by itself doesn't provide native storage so that needs to be taken care of by the host system & language (the browser for example).
+
+== Conclusion
+
+=== Native Rust vs bindings to C/C++
+
+I gave a `+` for all pure-Rust implementations because it makes the build process slightly easier.
+However, not being a pure-Rust library is not a show-stopper at all, hence I didn't record it as a `-` if a library is written in a different language (actually, most of them are).
+
+=== Type of storage backend
+
+I'd say there is generally little advantage of having a `Document`-based storage vs having a `Key/Value`-based one.
+The challenges in terms of integration will be roughly the same:
+
+- which collections to we want? (similar to deciding on a namespacing-strategy)
+- how do we want to serialize the data?
+
+Some `Document`-based solutions provide additional features though which might be useful.
+For example, `ejdb` has support for transactions
+
+An `SQL`-based backend would have the advantage that we could use `diesel` which would do the heavy lifting of crafting the correct SQL-statements for us.
+Additionally, it provides a type-safe interface for interacting with the database.
+It is also a fairly popular library in the Rust ecosystem (4k stars).
+
+=== Storing blocks for `btsieve`
+
+Blockchain blocks are binary blobs that contain no private information.
+Storing them is (for our usecase) mostly interesting for performance reasons so that we have them available after restarts of `btsieve`.
+They could easily also be stored on disks in regular files and would not have to go into a database.
+
+=== Recommendation
+
+Independently of the actual storage backend, it would very likely be smart to implement an extra layer of types behind the storage interface that we expose to our application core.
+This would allow us to evolve the domain model in the application's core in the way we see fit without directly affecting the storage model.
+
+We have to keep in mind that every change in the storage model's structure needs to be accomodated by a migration strategy once `cnd` is released and used by people.
+(Alternatively we may not make any guarantees about the data before a certain version i.e. 1.0 and always delete the database and start from scratch.
+In the end, only historical data will be lost, assuming the user doesn't upgrade COMIT whilest having an on-going SWAP.)
```

Comment:
It's a very good point that we may look into somehow enforcing (e.g. stopping migration DB if a swap is in progress)